src/framework/components/esmscript/component.js

```diff
@@ -0,0 +1,525 @@
+import { Debug } from '../../../core/debug.js';
+import { Component } from '../component.js';
+import { Entity } from '../../entity.js';
+
+/**
+ * @typedef {Object|Map} AttributeDefinition
+ * @property {'asset'|'boolean'|'curve'|'entity'|'json'|'number'|'rgb'|'rgba'|'string'|'vec2'|'vec3'|'vec4'} type - The attribute type
+ */
+
+/**
+ * @callback UpdateFunction
+ * @param {number} dt - The time since the last update.
+ */
+
+/**
+ * @callback SwapFunction
+ * @param {Object} newState - The new state to swap to.
+ */
+
+/**
+ * @typedef {Object} ModuleInstance
+ * @property {Function} [initialize] - A function called when the component is mounted
+ * @property {UpdateFunction} [update] - A function called on game tick if the module is enabled
+ * @property {Function} [destroy] - A function called when the module should be destroyed
+ * @property {SwapFunction} [swap] - A function called to swap state
+ */
+
+/**
+ * Checks if a class contains a method either itself or in it's inheritance chain
+ *
+ * @param {Function} Class - The class to check
+ * @param {string} method - The name of the method to check
+ * @returns {boolean} if a valid class and contains the method in it's inheritance chain
+ */
+const doesClassHaveMethod = (Class, method) => {
+    return typeof Class === 'function' &&
+        typeof Class.prototype === 'object' &&
+        method in Class.prototype;
+};
+
+/**
+ * The EsmScriptComponent extends the functionality of an Entity by
+ * allowing you to attach your own ESM modules to it.
+ *
+ * @augments Component
+ */
+class EsmScriptComponent extends Component {
+    /**
+     * A list of valid attribute types
+     */
+    static VALID_ATTR_TYPES = new Set([
+        "asset",
+        "boolean",
+        "curve",
+        "entity",
+        "json",
+        "number",
+        "rgb",
+        "rgba",
+        "string",
+        "vec2",
+        "vec3",
+        "vec4"
+    ]);
+
+    /**
+     * For any given attribute definition returns whether it's shape
+     * is similar to a valid attribute definition.
```

Comment:
what does it mean `similar`? The code does not seem to handle similarity.

Reply:
The comment is confusing, but basically if a object contains a valid `type` property, then it can be treated as an attribute.